src/Npgsql/NpgsqlSlimDataSourceBuilder.cs

```diff
@@ -276,6 +275,27 @@ internal void AddDefaultTypeResolverFactory(TypeHandlerResolverFactory resolverF
         throw new Exception("No built-in resolver factory found");
     }
 
+    void AddUserMappingResolver(IUserTypeMapping userMapping)
+    {
+        RemoveUserMappingResolver(userMapping.PgTypeName);
+        var factory = new UserMappedTypeHandlerResolverFactory(userMapping);
```

Comment:
Originally I thought of the resolvers list as always being quite small, e.g. how many plugins could there possibly be.. I'm not sure if we have places where perf would be problematic if this list starts growing; IIRC we should only actually access the resolvers for NpgsqlDbType/Type that we haven't seen before (we cache). Maybe on connection startup (when the cache is empty) this could in theory become problematic, if the user has e.g. 200 enums/composites or something.

To be on the safe side, we could have a single resolver handle them all; though if you're convinced this doesn't matter I'm OK with it.

Reply:
> Maybe on connection startup (when the cache is empty) this could in theory become problematic, if the user has e.g. 200 enums/composites or something.

You do have a point here. A few of my thoughts below:
1. With 200+ enums/composites a cold start will be slow no matter what we do (as it implies a big codebase).
2. One way to go around this is to implement resolver sorting. Since enums/composites don't interfere with BuiltInResolver (and the same thing can be said about all of the resolvers we implement in #4495) we put BuiltInResolver first and enums/composites last (+ resolvers from #4495).
3. While we do cache the results of the resolvers, there is an exception to that rule: resolving by value. Because of DateTime's kind we have to return either timestamp or timestamptz.

In light of the above I propose:
1. We do implement a single resolver for custom types.
2. We consider sorting resolvers (putting BuiltInResolver first, most of out internal resolvers after, and anything a user adds before BuiltInResolver).

What do you say?

Reply:
> We do implement a single resolver for custom types.

Though that's probably going to complicate things for the global type mapper...

Reply:
@NinoFloris what do you think of the above? We've discussed simplying our type handler concepts by bringing ideas across from Slon; if we do that, it may not be worth worrying too much about the current situation...

Reply:
Right, Slon's model also relies on resolvers for this work so in that respect nothing changes. If there are a huge number of (user) resolvers to check before ending up at the builtin set it would equally slow things down until the cache is properly filled. 

I would probably have something like the following order:
- Builtin
- User mappings (just one resolver)
- (Additive) plugins (however many resolvers as there are plugins) 

In Slon the only time you need another resolver to be first is for overriding the given converter for some clr type or overriding info (default clr type for a pg type, text writing preferred etc).

Reply:
Just throwing it out there :)

Now that type mapping is managed at the data source level, are we sure we actually need a list of resolvers? In other words, calling UseNodaTime could simply add the appropriate handlers directly into the type mapper. This means that NpgsqlDataSourceBuilder would hold a type mapper and progressively configure it as e.g. UseNodaTime is called, and finally just hand it off to the data source when Build() is called.

One of the reasons for the current design was to be lazy, i.e. not eagerly create handlers for all the built-in types etc. But that was from the era where we had a type mapper for each connection; if we have just one for the type mapper, I'm not sure we should care.

A slightly tricky thing would be the global type mapper, which we currently "inherit" in the data source. Doing calls on it could still just update a (regular) type mapper, which the data source builder could just copy when it's created.

This would be a big change, and I'm sure there would be things to work out; but do we want to consider it?

Reply:
I don't see how that works when things are per clr type, how would you be anything but lazy in that case?